cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,414 @@ struct ReplaceWithMaskFunctor {
+                                  last_valid_value_chunk, last_valid_value_offset);
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct FillNullExecutor<Type, enable_if_fixed_size_binary<Type>> {
+  static Status ExecFillNull(KernelContext* ctx, const ArrayData& array,
+                             const uint8_t* reversed_bitmap, ArrayData* output,
+                             int8_t direction, const ArrayData& last_valid_value_chunk,
+                             int64_t* last_valid_value_offset) {
+    FillNullInDirectionImpl<Type>(array, reversed_bitmap, output, direction,
+                                  last_valid_value_chunk, last_valid_value_offset);
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct FillNullExecutor<Type, enable_if_base_binary<Type>> {
+  using offset_type = typename Type::offset_type;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+
+  static Status ExecFillNull(KernelContext* ctx, const ArrayData& current_chunk,
+                             const uint8_t* reversed_bitmap, ArrayData* output,
+                             int8_t direction, const ArrayData& last_valid_value_chunk,
+                             int64_t* last_valid_value_offset) {
+    BuilderType builder(current_chunk.type, ctx->memory_pool());
+    RETURN_NOT_OK(builder.Reserve(current_chunk.length));
+    RETURN_NOT_OK(builder.ReserveData(current_chunk.buffers[2]->size()));
+    int64_t array_value_index = direction == 1 ? 0 : current_chunk.length - 1;
+    const uint8_t* data = current_chunk.buffers[2]->data();
+    const uint8_t* data_prev = last_valid_value_chunk.buffers[2]->data();
+    const offset_type* offsets = current_chunk.GetValues<offset_type>(1);
+    const offset_type* offsets_prev = last_valid_value_chunk.GetValues<offset_type>(1);
+
+    bool has_fill_value_last_chunk = *last_valid_value_offset != -1;
+    bool has_fill_value_current_chunk = false;
+    /*tuple for store: <use current_chunk(true) or last_valid_chunk(false),
+     * start offset of the current value, end offset for the current value>*/
+    std::vector<std::tuple<bool, int64_t, int64_t>> offsets_reversed;
+    RETURN_NOT_OK(VisitNullBitmapInline<>(
+        reversed_bitmap, output->offset, current_chunk.length,
+        current_chunk.GetNullCount(),
+        [&]() {
+          const offset_type offset0 = offsets[array_value_index];
+          const offset_type offset1 = offsets[array_value_index + 1];
+          offsets_reversed.push_back(
+              std::make_tuple(/*current_chunk=*/true, offset0, offset1 - offset0));
+          *last_valid_value_offset = array_value_index;
+          has_fill_value_current_chunk = true;
+          has_fill_value_last_chunk = false;
+          array_value_index += direction;
+          return Status::OK();
+        },
+        [&]() {
+          if (has_fill_value_current_chunk || has_fill_value_last_chunk) {
```

Comment:
nit: why not make this just one if-else if-else?

Reply:
Ping here.

Reply:
Ah sorry, I see the comments now.

For here: instead of 

```
if (A || B) {
  if (!B) // A
  else // B
} else {
  // C
}
```

why not

```
if (A) // A
else if (B) // B
else // C
```

?